airbyte-integrations/bases/debezium-v1-9-2/src/main/java/io/airbyte/integrations/debezium/internals/PostgresDebeziumStateUtil.java

```diff
@@ -0,0 +1,154 @@
+import io.debezium.connector.postgresql.connection.Lsn;
+import io.debezium.pipeline.spi.Offsets;
+import io.debezium.pipeline.spi.Partition;
+import java.util.Collections;
+import java.util.Map;
+import java.util.Optional;
+import java.util.OptionalLong;
+import java.util.Properties;
+import java.util.Set;
+import org.apache.kafka.connect.json.JsonConverter;
+import org.apache.kafka.connect.runtime.WorkerConfig;
+import org.apache.kafka.connect.runtime.standalone.StandaloneConfig;
+import org.apache.kafka.connect.storage.FileOffsetBackingStore;
+import org.apache.kafka.connect.storage.OffsetStorageReaderImpl;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This class is inspired by Debezium's Postgres connector internal implementation on how it parses
+ * the state
+ */
+public class PostgresDebeziumStateUtil {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(PostgresDebeziumStateUtil.class);
+
+  public boolean isSavedOffsetAfterReplicationSlotLSN(final Properties baseProperties,
+                                                      final ConfiguredAirbyteCatalog catalog,
+                                                      final JsonNode cdcState,
+                                                      final JsonNode replicationSlot,
+                                                      final JsonNode config) {
+
+    final DebeziumPropertiesManager debeziumPropertiesManager = new DebeziumPropertiesManager(baseProperties, config, catalog,
+        AirbyteFileOffsetBackingStore.initializeState(cdcState),
+        Optional.empty());
+    final Properties debeziumProperties = debeziumPropertiesManager.getDebeziumProperties();
+    final OptionalLong savedOffset = parseSavedOffset(debeziumProperties);
+
+    if (savedOffset.isPresent()) {
+      if (replicationSlot.has("confirmed_flush_lsn")) {
+        final long confirmedFlushLsnOnServerSide = Lsn.valueOf(replicationSlot.get("confirmed_flush_lsn").asText()).asLong();
+        LOGGER.info("Replication slot confirmed_flush_lsn : " + confirmedFlushLsnOnServerSide + " Saved offset LSN : " + savedOffset.getAsLong());
+        return savedOffset.getAsLong() >= confirmedFlushLsnOnServerSide;
+      } else if (replicationSlot.has("restart_lsn")) {
+        final long restartLsn = Lsn.valueOf(replicationSlot.get("restart_lsn").asText()).asLong();
+        LOGGER.info("Replication slot restart_lsn : " + restartLsn + " Saved offset LSN : " + savedOffset.getAsLong());
+        return savedOffset.getAsLong() >= restartLsn;
+      }
+    }
+
+    return true;
+  }
+
+  /**
+   *
+   * @param properties Properties should contain the relevant properties like path to the debezium
+   *        state file, etc. It's assumed that the state file is already initialised with the saved
+   *        state
+   * @return Returns the LSN that Airbyte has acknowledged in the source database server
+   */
+  private OptionalLong parseSavedOffset(final Properties properties) {
```

Comment:
I think the approach here makes sense, but it looks like we're depending on debezium internal behaviors. Are we putting ourselves at risk of really complicated debezium version upgrades? (I don't think it's avoidable, just want to check that we're OK with it)

Reply:
As long as debezium doesnt introduce a breaking change in the structure of the state, we should be fine. We upgraded from 1.4 to 1.9 and there was no breaking change in the structure of the state that debezium stores so this logic would have stayed the same in that upgrade.